Initial implementation of Tikhonov regularization

mguddati opened the pull request:
Initial implementation of Tikhonov regularization and add a test

(Hessian needs to be done still, and test with taonls) 
(Need to add documentation to MD file)
(Maybe move regularization from executioner to optimization reporter)

ref #24101

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->

## Design
<!--A concise description (design) of the enhancement.-->

## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

mguddati committed:
Initial implementation of Tikhonov regularization and associate test
(Hessian needs to be done still, and testd with taonls)
(Need to add documentation to MD file)
(Maybe move regularization from executioner to optimization reporter)

ref #24101

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1467541/) on b3bf8ad wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24102/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24102/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1542ca05a186584a315c5643fec9329c9d2b5478`

mguddati committed:
Added hessian part for Tikhonov regularization
Changed the test to NLS to ensure that the method
converges immediately and to the same solution
irrespective of the starting point

ref #24101

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1467654/) on 98732fa wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24102/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24102/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1542ca05a186584a315c5643fec9329c9d2b5478`

mguddati committed:
Cleanup format

ref #24102

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1467758/) on 27665b9 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24102/site)

This comment will be updated on new commits.

lynnmunday commented:
This should be closed.  I pulled @mguddati pr, made some changes opened this PR. https://github.com/idaholab/moose/pull/25080

lynnmunday commented:
@loganharbour will you close this PR, I pulled it and merged it into a different PR https://github.com/idaholab/moose/pull/25080
